### DISCUSSION TITLE
How to include code snippet from external file?

### DISCUSSION THREAD

**Discussion by drriguz** [quarto-dev/quarto-cli#3868] [Q&A] [answered]
Hi,
I'm trying display only certain lines of code from source file, eg:
![image](https://user-images.githubusercontent.com/6427851/210953079-2f593c1d-4888-4e70-8f0f-bc07362eac25.png)

I can include the full source code like bellow, however I cannot find an option to set line ranges, is't supported?
```groovy
{{< include code/ch-01/jenkins-example.groovy >}}
```

**Comment by mcanouil** [comment, marked as answer]
Include shortcode is to include  child Quarto document.

For actual code, via `knitr` at least you can use <https://quarto.org/docs/reference/cells/cells-knitr.html#include>.

Meaning, there is no native way to include only part of a file, unless you write a hacky code cell using the knitr `code` option to include your contents.

In my opinion, if you only want a subset of a Quarto document A (not code) into document B, I suggest you put this subset in _C, and include it in both A and B.

**Reply by drriguz** [reply]
Thanks! My purpose is mainly to include part of long code in the document, which would be useful when writing a book.
So it seems that it could be implemented by writing some code with knitr, I'll have a try!

**Reply by cderv** [reply]
There is also an extension now to do this : https://github.com/quarto-ext/include-code-files

You could find that useful

**Reply by drriguz** [reply]
Thanks! That's exactlly what I'm looking for. I was trying  [b3/include-code-files](https://github.com/b3/include-code-files) and it's glad to see that the extension has been included into the official repository.

**Comment by fuhrmanator** [comment]
Just an idea for a workaround: put a link to the source on GitHub, if the code is there, such as https://github.com/fuhrmanator/pandoc-filter-fr-nbsp/blob/main/fr-nbsp.lua#L37-L40  -- that's `https://github.com/fuhrmanator/pandoc-filter-fr-nbsp/blob/main/fr-nbsp.lua#L37-L40`

I don't think you can display the GitHub view the same way in the book, but maybe you can put a live link (as above) and put a screen shot in the book? It's probably best to reference the code at a specific commit since the latest version might change.

**Reply by drriguz** [reply]
A link is more suitable if it's a e-book, however I'm writing a paper book, so still need to display it directly.

**Comment by DaniMori** [comment]
FWIW, I was able to do this with `code:` (it was easier than I expected in fact). Here's my example:

````
```{r}
#| code: !expr readLines("R/simulated_data.R")[20:39]
```
````

It shows lines 20 to 39 of the script file in the output, as expected.
